### DISCUSSION TITLE
Change HTTP2 max concurrent streams value

### DISCUSSION THREAD

**Discussion by dunkmann00** [encode/httpx#2416] [Q&A] [answered]
Is it possible to change the max concurrent streams value for an HTTP2 connection? I was looking into this but it appears that the default value from the [H2Connection](https://github.com/python-hyper/h2/blob/63b6b97e691334464a57f933bfd7ed444f364708/src/h2/connection.py#L324) class is used in the [HTTP2Connection](https://github.com/encode/httpcore/blob/d3a847c14981de7c697b6b71c82955e38f0a3344/httpcore/_async/http2.py#L52) object, and cannot be changed. Is this intended to be not configurable, or am I missing something?

Thanks!

Note: I had originally opened an issue in encode/httpcore/issues/598 but if this is the more appropriate forum to discuss this, I will close that.

**Comment by lovelydinosaur** [comment, marked as answer]
> Is it possible to change the max concurrent streams value for an HTTP2 connection?

Currently, no.

**Reply by lovelydinosaur** [reply]
@JB0925 also brings this up on gitter...

https://gitter.im/encode/community?at=63684916cf41c67a5ce7d152

We don't currently support altering the `h2` settings at all.

https://github.com/encode/httpcore/blob/db0d4e1d46a630240fdf260ab99e9c75d9cd7c7a/httpcore/_async/http2.py#L156-L166

**Reply by dunkmann00** [reply]
Are you open to `httpx` supporting the ability to alter `h2` settings?

**Reply by lovelydinosaur** [reply]
Maybe. Having a demonstrated use-case would be a good starting point.
Do you want to turn the max concurrent streams value up, or down? Why? Can you demo that a different value works better for you in some way?

**Reply by dunkmann00** [reply]
I am trying to update a library that helps with sending Apple Push Notifications, [PyAPNs2](https://github.com/Pr0Ger/PyAPNs2). It was written with [hyper](https://github.com/python-hyper/hyper) as a dependency, but since that is no longer maintained (for quite some time as you are probably aware) I figured I would try replacing it with `httpx`.

I've gotten most of this working but had to stop at the part about handling multiple streams. I know this has come up before, but I'll include a snippet from Apple's docs about concurrent streams:
> APNs allows multiple concurrent streams for each connection, but don’t assume a specific number of streams. The exact number varies based on server load and whether you use a provider certificate or an authentication token. For example, when using an authentication token, APNs allows only one stream until you post a request with a valid authentication token. APNs ignores HTTP/2 PRIORITY frames, so don’t send them on your streams.

Source: [Sending Notification Requests to APNs](https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/sending_notification_requests_to_apns)

1) So to work with this, I could initially just set max concurrent streams to `1`, and then after sending a single request, check the remote value, and assuming it has changed to greater than `1`, update my local value to that.

2) Unless changing my local value because the remote value changed is not a good pattern. In this case it would be nice to set the local value to some higher value (say `1000`), and then `httpx` could determine how many concurrent streams are allowed based on both the `remote` and `local` max concurrent streams values (i.e. it would only send at most, the minimum of the two values).

3) If that is not something that `httpx` is going to handle, then it would be nice to at least raise the `local` max concurrent streams value, and then I could manually, in the `PyAPNs2` library, make sure to only send up to the `remote`'s max concurrent streams value.

I think for my use case (and maybe a general case?) it would be great if `httpx` just handled max concurrent streams like I described in 2. That way I could set the `local` max concurrent streams to some value I'm comfortable with as an upper bound, and `httpx` will figure out how many streams it can send concurrently based on checking the `remote` max concurrent streams value. In this case, I think its important to note you would have to support checking for changes to the `remote` max concurrent streams.

Hopefully this is clear and coherent...I did my best! Thoughts?

**Reply by lovelydinosaur** [reply]
Okay. It looks to me like "can I change the max concurrent streams setting" isn't *necessarily* actually what's needed here...

I'm up for spending some time working through this.

> had to stop at the part about handling multiple streams

Can you be more specific here...

* Do you see errors from the server, and if so what?
* Are you able to observe the HTTP/2 frames in each direction? (The ["trace" extension](https://www.encode.io/httpcore/extensions/#trace) is useful here)

There's more that could be said here, but let's try to take this step-by-step.

**Reply by dunkmann00** [reply]
> I'm up for spending some time working through this.

I appreciate that, thank you!

> had to stop at the part about handling multiple streams

I just meant there was no way for me to observe the remote max concurrent streams value (through the public parts of `httpx` anyway) and no way to change the local max concurrent streams value. So as far as updating `PyAPNs2`, I got it working with sending one stream at a time and *had to stop*. Although, I would like to be able to have it support sending up-to as many streams as the `APNs` server is willing to accept, which is how it worked with `hyper`. But I do not know what the remote server is willing to accept because that is not available in `httpx`, and even if I did know, there is currently no way to use that to change how many concurrent streams `httpx` will use for a connection.

> Okay. It looks to me like "can I change the max concurrent streams setting" isn't necessarily actually what's needed here...

I'm not sure I understand what you mean?

**Reply by lovelydinosaur** [reply]
> So as far as updating PyAPNs2, I got it working with sending one stream at a time and had to stop.

Okay, so the docs state "For example, when using an authentication token, APNs allows only one stream until you post a request with a valid authentication token"...

* Do *all* of your outgoing requests include an authentication token?
* What behaviour do you see if you attempt to send multiple concurrent requests?

Ideally you shouldn't need to be doing *any* stream concurrency on your side of the client at all, because `httpx` ought to be just handling that for you, establishing multiple connections if needed. I can see where there's a couple of possible gaps in how that handling currently works, but it'd be really helpful to observe any *actual* failure cases against a live server...

If you're able to hop onto a chat on https://gitter.im/encode/community sometime, we could see about working through this together.

**Reply by dunkmann00** [reply]
> Do all of your outgoing requests include an authentication token?

In my testing, yes. The other option, rather than using an auth token, is to use a certificate. But you have to use one or the other and I've just been sticking with the auth token method as that is what I have configured with Apple.

> What behaviour do you see if you attempt to send multiple concurrent requests?

I actually hadn't tried this until now. But the error I am getting when I tried to send 100 notifications concurrently:

`httpcore.LocalProtocolError: Max outbound streams is 1, 77 open` from [here](https://github.com/encode/httpcore/blob/23cbfca76a1aaad186e8fba1baf948506446d43d/httpcore/_async/http2.py#L142)

which was raised because of:

`h2.exceptions.TooManyStreamsError: Max outbound streams is 1, 77 open` from [here](https://github.com/python-hyper/h2/blob/63b6b97e691334464a57f933bfd7ed444f364708/src/h2/connection.py#L761-L764) in `h2`

Here is an excerpt from the trace log (thanks for pointing trace out btw):
```
...
http2.send_request_headers.started {'request': <Request [b'POST']>, 'stream_id': 149}
http2.send_request_headers.complete {'return_value': None}
http2.send_request_body.started {'request': <Request [b'POST']>, 'stream_id': 63}
http2.send_request_headers.started {'request': <Request [b'POST']>, 'stream_id': 151}
http2.send_request_headers.started {'request': <Request [b'POST']>, 'stream_id': 153}
http2.send_request_headers.started {'request': <Request [b'POST']>, 'stream_id': 155}
http2.send_request_headers.failed {'exception': TooManyStreamsError('Max outbound streams is 1, 77 open')}
http2.response_closed.started {'stream_id': 155}
http2.response_closed.complete {'return_value': None}
http2.send_request_headers.complete {'return_value': None}
http2.send_request_body.started {'request': <Request [b'POST']>, 'stream_id': 65}
http2.send_request_headers.started {'request': <Request [b'POST']>, 'stream_id': 155}
http2.send_request_headers.failed {'exception': TooManyStreamsError('Max outbound streams is 1, 77 open')}
http2.response_closed.started {'stream_id': 155}
http2.send_request_headers.started {'request': <Request [b'POST']>, 'stream_id': 155}
http2.send_request_headers.failed {'exception': TooManyStreamsError('Max outbound streams is 1, 77 open')}
http2.response_closed.started {'stream_id': 155}
http2.send_request_body.complete {'return_value': None}
http2.receive_response_headers.started {'request': <Request [b'POST']>, 'stream_id': 11}
http2.response_closed.complete {'return_value': None}
http2.send_request_headers.started {'request': <Request [b'POST']>, 'stream_id': 155}
http2.send_request_headers.failed {'exception': TooManyStreamsError('Max outbound streams is 1, 77 open')}
http2.response_closed.started {'stream_id': 155}
http2.send_request_headers.failed {'exception': CancelledError()}
http2.send_request_body.failed {'exception': CancelledError()}
...
```

Afterwards it is just more `CancelledError()`. It does seem like a bunch will complete, not sure why `h2` doesn't stop it right away. I tried it a second time and it failed at `48 open` rather than `77` so I don't know what's going on there, guess it has something to do with the order the coroutines end up running?

> Ideally you shouldn't need to be doing any stream concurrency on your side of the client at all, because httpx ought to be just handling that for you, establishing multiple connections if needed.

That makes sense to me. Originally I wasn't sure it was something `httpx` wanted to support, which is why I was trying to figure out how to handle it manually, but that being something invisible to the end user seems great.

Chatting on gitter sounds good. I should have time tomorrow so I'll try to be on as much as I can.

**Comment by tinycogio** [comment]
Hello!
Was there any progress on this issue? Like dunkman00, I'm trying to migrate to httpx for sending APNs and currently running into some perf issues with the max number of streams set to 1. I can try to play around with it more on our next push blast.

Thanks!

**Reply by dunkmann00** [reply]
Hello @tinycogio! Tom & I discussed this some more but I have unfortunately not gotten a chance to work on this. I should have time this week and plan to get back to this.

I will post back here when I have made some progress.

**Comment by dunkmann00** [comment]
@tomchristie Whenever you get a chance, I have opened https://github.com/encode/httpcore/pull/652 to adrdess this issue.
